File: src/ccpp_ipd.f90
!>
!! @brief The IPD module.
!!
!! @details The IPD routines for calling the specified
!!          physics schemes.
!
module ccpp_ipd

    use, intrinsic :: iso_c_binding,                                   &
                      only: c_int32_t, c_char, c_ptr, c_loc, c_funptr
    use            :: ccpp_types,                                      &
                      only: ccpp_t, ccpp_scheme_t
    use            :: ccpp_errors,                                     &
                      only: ccpp_error
    use            :: ccpp_strings,                                    &
                      only: ccpp_cstr

    implicit none

    private
    public :: ccpp_ipd_init,                                           &
              ccpp_ipd_fini,                                           &
              ccpp_ipd_run

    interface
        integer(c_int32_t)                                             &
        function ccpp_ipd_open                                         &
                   (name, library, version, shdl, lhdl)                &
                   bind(c, name='ccpp_ipd_open')
            import :: c_char, c_int32_t, c_ptr, c_funptr
            character(kind=c_char), dimension(*)  :: name
            character(kind=c_char), dimension(*)  :: library
            character(kind=c_char), dimension(*)  :: version
            type(c_ptr)                           :: shdl
            type(c_ptr)                           :: lhdl
        end function ccpp_ipd_open

        integer(c_int32_t)                                             &
        function ccpp_ipd_close                                        &
                   (lhdl)                                              &
                   bind(c, name='ccpp_ipd_close')
            import :: c_int32_t, c_ptr, c_funptr
            type(c_ptr)                           :: lhdl
        end function ccpp_ipd_close

        integer(c_int32_t)                                             &
        function ccpp_ipd_cap                                          &
                   (shdl, cdata)                                       &
                   bind(c, name='ccpp_ipd_cap')
            import :: c_int32_t, c_ptr, c_funptr
            type(c_ptr)                           :: shdl
            type(c_ptr)                           :: cdata
        end function ccpp_ipd_cap
    end interface

    contains

    !>
    !! IPD initialization subroutine.
    !!
    !! This loops over all defined schemes to obtain
    !! a function pointer to them.
    !!
    !! @param[in,out] cdata    The CCPP data of type ccpp_t
    !! @param[  out]  ierr     Integer error flag
    !
    subroutine ccpp_ipd_init(cdata, ierr)

        type(ccpp_t), target, intent(inout)  :: cdata

        integer                              :: ierr
        integer                              :: i
        integer                              :: j
        integer                              :: k

        ierr = 0

        do i=1, cdata%suite%ipds_max
            do j=1, cdata%suite%ipds(i)%subcycles_max
                do k=1, cdata%suite%ipds(i)%subcycles(j)%schemes_max

                    associate (s => cdata%suite%ipds(i)%subcycles(j)%schemes(k))
                    ierr = ccpp_ipd_open(ccpp_cstr(s%name),    &
                                         ccpp_cstr(s%library), &
                                         ccpp_cstr(s%version), &
                                         s%scheme_hdl,         &
                                         s%library_hdl)
                    if (ierr /= 0) then
                        call ccpp_error('A problem occured loading ' &
                                        // trim(s%name) // ' from '  &
                                        // trim(s%library))
                    end if
                    end associate

                end do
            end do
        end do

    end subroutine ccpp_ipd_init

    !>
    !! IPD finalization subroutine.
    !!
    !! This loops over all defined schemes to close
    !! the handle to the scheme library
    !!
    !! @param[in,out] cdata    The CCPP data of type ccpp_t
    !! @param[   out] ierr     Integer error flag
    !
    subroutine ccpp_ipd_fini(cdata, ierr)

        type(ccpp_t), target, intent(inout)  :: cdata

        integer                              :: ierr
        integer                              :: i
        integer                              :: j
        integer                              :: k

        ierr = 0

        do i=1, cdata%suite%ipds_max
            do j=1, cdata%suite%ipds(i)%subcycles_max
                do k=1, cdata%suite%ipds(i)%subcycles(j)%schemes_max

                    associate (s => cdata%suite%ipds(i)%subcycles(j)%schemes(k))
                    ierr = ccpp_ipd_close(s%library_hdl)
                    if (ierr /= 0) then
                        call ccpp_error('A problem occured close ' &
                                        // trim(s%library))
                    end if
                    end associate
                end do
            end do
        end do

    end subroutine ccpp_ipd_fini

    !>
    !! The run subroutine for the IPD. This will call
    !! the various physics schemes as specified in the
    !! suite data.
    !!
    !! @param[in    ] scheme  The scheme to run
    !! @param[in,out] cdata   The CCPP data of type ccpp_t
    !! @param[   out] ierr     Integer error flag
    !
    subroutine ccpp_ipd_run(scheme, cdata, ierr)

        type(ccpp_scheme_t),  intent(in   )  :: scheme
        type(ccpp_t), target, intent(inout)  :: cdata
        integer,              intent(  out)  :: ierr

        ierr = 0

        ierr = ccpp_ipd_cap(scheme%scheme_hdl, c_loc(cdata))
        if (ierr /= 0) then
            call ccpp_error('A problem occured calling '// &
                            trim(scheme%name) //' scheme')
        end if

    end subroutine ccpp_ipd_run

end module ccpp_ipd

File: src/ccpp_ipd.c
/**
 * @file ccpp_ipd.c
 *
 * Routines and functions for the IPD to call physics routines.
 *
 * @ingroup IPD
 * @{
 **/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <dlfcn.h>
#include <err.h>
#include <sysexits.h>

#include "ccpp_ipd.h"

/** Shared library prefix and suffix for different platforms **/
static const char prefix[] = "lib";
#if __APPLE__
static const char suffix[] = ".dylib";
#elif __unix__
static const char suffix[] = ".so";
#endif

/**
 * IPD cap initialization routine.
 *
 * @param[in]  scheme    The scheme name to call.
 * @param[in]  lib       The library continaing the physics scheme.
 * @param[in]  ver       The library version number.
 * @param[out] shdl      The scheme function pointer handle.
 * @param[out] lhdl      The library handle.
 * @retval     0         If it was sucessful
 * @retval     1         If there was an error
 **/
int
ccpp_ipd_open(const char *scheme, const char *lib, const char *ver,
	      void **shdl, void **lhdl)
{
	int i = 0;
	int n = 0;
	const char cap[] = "_cap";
	char *library = NULL;
	char *scheme_cap = NULL;
	char *error = NULL;

	/* Generate the library name with the platform suffix */
	n = (strlen(prefix) + strlen(lib) + strlen(suffix) + strlen(ver) +2)
		*sizeof(char);
	library = malloc(n);
	memset(library, 0, n);
	if (strcmp(ver, "") != 0) {
		snprintf(library, n, "%s%s.%s%s", prefix, lib, ver, suffix);
	} else {
		snprintf(library, n, "%s%s%s", prefix, lib, suffix);
	}

	/* Generate the scheme cap function name */
	n = (strlen(scheme) +strlen(cap) +1)*sizeof(char);
	scheme_cap = malloc(n);
	memset(scheme_cap, 0, n);

	for (i=0; i < n; ++i) {
		scheme_cap[i] = tolower(scheme[i]);
	}

	strncat(scheme_cap, cap, n);

	/* Open a handle to the library */
	*lhdl = dlopen(library, RTLD_NOW);
	if (!*lhdl) {
		warnx("%s", dlerror());
		return(EXIT_FAILURE);
	}

	dlerror();
	/*
	*(void **)(&(*shdl)) = dlsym(*lhdl, scheme_cap);
	*/
	*(void **)shdl = dlsym(*lhdl, scheme_cap);
	if ((error = dlerror()) != NULL)  {
		warnx("%s", error);
		return(EXIT_FAILURE);
	}

	/* Free the library filename */
	if (library) {
		free(library);
		library = NULL;
	}

	/* Free the scheme cap function name */
	if (scheme_cap) {
		free(scheme_cap);
		scheme_cap = NULL;
	}

	return(EXIT_SUCCESS);
}

/**
 * IPD library finialization routine.
 *
 * @param[in] lhdl      The library handle.
 * @retval     0        If it was sucessful
 * @retval     1        If there was an error
 **/
int
ccpp_ipd_close(void **lhdl)
{
	char *error = NULL;

	dlerror();
	dlclose(lhdl);
	if ((error = dlerror()) != NULL)  {
		warnx("%s", error);
		return(EXIT_FAILURE);
	}

	return(EXIT_SUCCESS);
}

/**
 * IPD cap calling routine.
 *
 * @param[in] f_ptr     The scheme function pointer to call.
 * @param[in] data      The opaque ccpp_t data type to pass.
 * @retval     0        If it was sucessful
 * @retval     1        If there was an error
 **/
int
ccpp_ipd_cap(void **f_ptr, void **data)
{
	void (*fun)(void **) = *f_ptr;

	/* Call the schemes cap subroutine */
	fun(data);

	return(EXIT_SUCCESS);
}

/**
 * @}
 **/
